ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -15,17 +15,175 @@
 module D = Debug.Make(struct let name="xapi_ha_vm_failover" end)
 open D
 
+type task_result = (Rpc.t, exn) Result.result
+
+(** [result_of_task ~__context task] returns the status of [task]
+ * unless it is still pending. Exceptions are converted into [Result.Error] *)
+let result_of_task ~__context self : task_result option =
+  try match Db.Task.get_status ~__context ~self with
+    | `pending -> None
+    | `success ->
+      Some (Result.Ok (match Db.Task.get_result ~__context ~self with
+          | "" -> Rpc.Null
+          | s -> Xmlrpc.of_string s))
+    | `cancelled -> Some (Result.Error (Failure "Cancelled"))
+    | `cancelling -> Some (Result.Error (Failure "Cancelling"))
+    | `failure ->
+      Some (match Db.Task.get_error_info ~__context ~self with
+          | [] -> Result.Error (Failure "Unknown error")
+          | code :: params -> Result.Error (Api_errors.Server_error (code, params)))
+  with e ->
+    (* cannot fetch task status, maybe task got destroyed *)
+    Backtrace.is_important e;
+    Some (Result.Error e)
+
+
+module TaskChains : sig
+  (** the type of delayed task related actions *)
+  type +'a t
+
+  val return : 'a -> 'a t
+
+  val ok : 'a -> ('a, exn) Result.result t
+
+  val fail : exn -> ('a, exn) Result.result t
+
+  val task : (unit -> API.ref_task) -> task_result t
+  (**[task f] is an action that evaluates to the result of task [f ()] *)
+
+  val fmap : ('a -> 'b) -> 'a t -> 'b t
+  (** [fmap f t] will map the final result of action [t] through the function [f].
+     * It is not guaranteed that [f] is executed as soon as the result of [t] is available. *)
+
+  (** this module is meant to be opened, defines only an infix operator,
+   * avoids polluting the namespace with other functions *)
+  module Infix : sig
+    val ( >>= ) : 'a t -> ('a -> 'b t) -> 'b t
+    (** [m >>= k] monadic bind: executes [k] with the result of [m] computation.
+     * It is not guaranteed that [k] is executed as soon as the result of [m] is available.
+     * *)
+  end
+
+  val eval : __context:Context.t -> 'a t -> 'a t
+  (** [eval ~__context t] will either return [t] unchanged,
+      or evaluate the next action after receiving the result of the contained task.
+      If the action raises an exception, that exceptions propagates out of this function *)
+
+  val parallel :
+    __context:Context.t -> rpc:(Rpc.call -> Rpc.response)
+    -> session_id:API.ref_session -> ('a * ('b, exn) Result.result t) list
+    -> ('a * ('b, exn) Result.result) list
+  (** [parallel ~__context lst] groups actions in [lst] by order number and
+   * executes all actions with same order number in parallel.
```

Comment:
This is actually done in map_parallel, this comment ended up in the wrong place.